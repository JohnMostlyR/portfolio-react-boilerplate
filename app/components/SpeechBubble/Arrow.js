import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { lighten } from 'polished';

/* eslint-disable indent */
const SVG = styled.svg.attrs({
  xmlns: 'http://www.w3.org/2000/svg',
  viewBox: '0 0 848 640',
  preserveAspectRatio: 'xMinYMid meet',
  focusable: 'false',
})`
  display: inline-block;
  align-self: ${({ arrowPosition }) =>
    arrowPosition.endsWith('-left') ? 'flex-start' : 'flex-end'};
  height: 1px;
  margin: -1px 0 0 0 !important;
  transform: rotateX(
      ${({ arrowPosition }) =>
        arrowPosition.startsWith('bottom-') ? '0deg' : '180deg'}
    )
    rotateY(
      ${({ arrowPosition }) =>
        arrowPosition.endsWith('-left') ? '0deg' : '180deg'}
    );
  z-index: 9;

  &::after {
    content: '' !important;
    display: block !important;
    clear: both !important;
  }

  @media (min-height: ${({ showBreakpoint, height }) =>
      showBreakpoint || height}) {
    height: ${({ height }) => height || '100%'};
  }
`;

function Arrow({
  arrowPosition,
  backgroundColor,
  height,
  isGhost,
  showBreakpoint,
}) {
  const lighterBackgroundColor =
    (backgroundColor && lighten(0.1, backgroundColor)) || 'inherit';

  return (
    <SVG
      arrowPosition={arrowPosition}
      height={height}
      showBreakpoint={showBreakpoint}
    >
      <path
        stroke={isGhost ? backgroundColor : 'none'}
        strokeWidth="24"
        fill={isGhost ? 'transparent' : lighterBackgroundColor}
        d="M424.2 424.4L526.9 640V398.7"
      />
      <path
        stroke={isGhost ? backgroundColor : 'none'}
        strokeWidth="24"
        fill={isGhost ? 'transparent' : backgroundColor}
        d="M848 318.3L424.2 424.4V158.9"
      />
      <path
        stroke={isGhost ? backgroundColor : 'none'}
        strokeWidth="24"
        fill={isGhost ? 'transparent' : lighterBackgroundColor}
        d="M0 0l848 318.3L636.4 0H0z"
      />
    </SVG>
  );
}

Arrow.propTypes = {
  arrowPosition: PropTypes.oneOf([
    'bottom-left',
    'bottom-right',
    'top-left',
    'top-right',
  ]),
  backgroundColor: PropTypes.string,
  height: PropTypes.string,
  isGhost: PropTypes.bool,
  showBreakpoint: PropTypes.string,
};

export default Arrow;
