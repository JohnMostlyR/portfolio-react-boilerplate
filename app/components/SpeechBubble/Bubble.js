import styled from 'styled-components';
import PropTypes from 'prop-types';

/* eslint-disable indent */
const Bubble = styled.div`
  display: inline-block;
  margin: 0 !important;
  background-color: ${({ backgroundColor, isGhost }) =>
    isGhost ? 'transparent' : backgroundColor || 'transparent'};
  border: ${({ backgroundColor = 'inherit', isGhost }) =>
    isGhost ? `2px solid ${backgroundColor}` : 'none'};
  border-radius: 0.5rem;
  box-shadow: 0 0 10px grey;
  color: ${({ color }) => color || 'inherit'};
  padding: ${({ padding }) => padding || 0};

  @media (min-height: ${({ showArrowBreakpoint }) =>
      showArrowBreakpoint || '100%'}) {
    margin-right: 0;
    border-bottom-left-radius: ${({ arrowPosition }) =>
      arrowPosition === 'bottom-left' ? 0 : '0.5rem'};
    border-bottom-right-radius: ${({ arrowPosition }) =>
      arrowPosition === 'bottom-right' ? 0 : '0.5rem'};
    border-top-left-radius: ${({ arrowPosition }) =>
      arrowPosition === 'top-left' ? 0 : '0.5rem'};
    border-top-right-radius: ${({ arrowPosition }) =>
      arrowPosition === 'top-right' ? 0 : '0.5rem'};
    box-shadow: none;
  }
`;

Bubble.propTypes = {
  arrowPosition: PropTypes.oneOf([
    'bottom-left',
    'bottom-right',
    'top-left',
    'top-right',
  ]),
  backgroundColor: PropTypes.string,
  color: PropTypes.string,
  isGhost: PropTypes.bool,
  padding: PropTypes.string,
  showArrowBreakpoint: PropTypes.string,
};

export default Bubble;
