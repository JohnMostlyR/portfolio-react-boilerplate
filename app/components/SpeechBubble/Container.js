import styled, { keyframes } from 'styled-components';
import PropTypes from 'prop-types';

const fadeIn = keyframes`
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
`;

/* eslint-disable indent */
const Container = styled.div`
  display: inline-flex;
  flex-flow: column nowrap;
  z-index: 10;
  filter: none;
  opacity: ${({ makeAppear }) => (makeAppear ? 0 : 1)};
  animation-delay: 0ms;
  animation-direction: normal;
  animation-duration: 1s;
  animation-fill-mode: both;
  animation-name: ${({ makeAppear }) => (makeAppear ? fadeIn : 'none')};
  animation-timing-function: ease-in-out;

  @media (min-height: ${({ showArrowBreakpoint }) =>
      showArrowBreakpoint || '100%'}) {
    filter: drop-shadow(0 0 5px grey);
  }

  @media (min-width: ${({ maxWidth }) => maxWidth || 'unset'}) {
    width: ${({ maxWidth }) => maxWidth || 'unset'};
  }
`;

Container.propTypes = {
  makeAppear: PropTypes.bool,
  maxWidth: PropTypes.string,
  showArrowBreakpoint: PropTypes.string,
};

export default Container;
