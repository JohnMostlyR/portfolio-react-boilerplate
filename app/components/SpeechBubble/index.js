import React from 'react';
import PropTypes from 'prop-types';

import Arrow from './Arrow';
import Bubble from './Bubble';
import Container from './Container';

class SpeechBubble extends React.PureComponent {
  render() {
    const {
      arrowHeight,
      arrowPosition,
      backgroundColor,
      color,
      children,
      isGhost,
      makeAppear,
      maxWidth,
      padding,
      showArrowBreakpoint,
    } = this.props;

    return (
      <Container
        showArrowBreakpoint={showArrowBreakpoint}
        maxWidth={maxWidth}
        makeAppear={makeAppear}
      >
        {arrowPosition.startsWith('top-') && (
          <Arrow
            arrowPosition={arrowPosition}
            backgroundColor={backgroundColor}
            height={arrowHeight}
            isGhost={isGhost}
            showBreakpoint={showArrowBreakpoint}
          />
        )}
        <Bubble
          arrowPosition={arrowPosition}
          backgroundColor={backgroundColor}
          color={color}
          isGhost={isGhost}
          padding={padding}
          showArrowBreakpoint={showArrowBreakpoint}
        >
          {children}
        </Bubble>
        {arrowPosition.startsWith('bottom-') && (
          <Arrow
            arrowPosition={arrowPosition}
            backgroundColor={backgroundColor}
            height={arrowHeight}
            isGhost={isGhost}
            showBreakpoint={showArrowBreakpoint}
          />
        )}
      </Container>
    );
  }
}

SpeechBubble.propTypes = {
  arrowHeight: PropTypes.string,
  arrowPosition: PropTypes.oneOf([
    'bottom-left',
    'bottom-right',
    'top-left',
    'top-right',
  ]),
  backgroundColor: PropTypes.string,
  color: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  isGhost: PropTypes.bool,
  makeAppear: PropTypes.bool,
  maxWidth: PropTypes.string,
  padding: PropTypes.string,
  showArrowBreakpoint: PropTypes.string,
};

SpeechBubble.defaultProps = {
  arrowPosition: 'bottom-left',
  isGhost: false,
};

export default SpeechBubble;
