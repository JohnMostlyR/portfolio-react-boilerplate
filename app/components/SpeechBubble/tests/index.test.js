import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import SpeechBubble from '../index';

describe('<SpeechBubble />', () => {
  it('should render and match the snapshot', () => {
    const renderedComponent = shallow(<SpeechBubble />);
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });

  it('should adopt the "arrowHeight" prop', () => {
    const renderedComponent = shallow(<SpeechBubble arrowHeight="1vw" />);
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });

  it('should adopt the "backgroundColor" prop', () => {
    const renderedComponent = shallow(<SpeechBubble backgroundColor="#000" />);
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });

  it('should adopt the "color" prop', () => {
    const renderedComponent = shallow(<SpeechBubble color="#fff" />);
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });

  it('should adopt the "maxWidth" prop', () => {
    const renderedComponent = shallow(<SpeechBubble maxWidth="10em" />);
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });

  it('should adopt the "padding" prop', () => {
    const renderedComponent = shallow(<SpeechBubble padding="0 0px" />);
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });

  it('should adopt the "showArrowBreakpoint" prop', () => {
    const renderedComponent = shallow(
      <SpeechBubble showArrowBreakpoint="100px" />,
    );
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });

  it('should adopt any children', () => {
    const renderedComponent = shallow(
      <SpeechBubble>
        <span>A Child</span>
      </SpeechBubble>,
    );
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });

  describe('the "isGhost" property', () => {
    it('should accept boolean "true" as value', () => {
      const renderedComponent = shallow(<SpeechBubble isGhost />);
      expect(toJson(renderedComponent)).toMatchSnapshot();
    });

    it('should accept boolean "false" as value', () => {
      const renderedComponent = shallow(<SpeechBubble isGhost={false} />);
      expect(toJson(renderedComponent)).toMatchSnapshot();
    });
  });

  describe('the "makeAppear" property', () => {
    it('should accept boolean "true" as value', () => {
      const renderedComponent = shallow(<SpeechBubble makeAppear />);
      expect(toJson(renderedComponent)).toMatchSnapshot();
    });

    it('should accept boolean "false" as value', () => {
      const renderedComponent = shallow(<SpeechBubble makeAppear={false} />);
      expect(toJson(renderedComponent)).toMatchSnapshot();
    });
  });

  describe('the arrowPosition property', () => {
    const renderedComponent = shallow(<SpeechBubble />);

    it('should accept "top-left" as value', () => {
      const fixture = 'top-left';
      renderedComponent.setProps({ arrowPosition: fixture });
      expect(toJson(renderedComponent)).toMatchSnapshot();
    });

    it('should accept "top-right" as value', () => {
      const fixture = 'top-right';
      renderedComponent.setProps({ arrowPosition: fixture });
      expect(toJson(renderedComponent)).toMatchSnapshot();
    });

    it('should accept "bottom-right" as value', () => {
      const fixture = 'bottom-right';
      renderedComponent.setProps({ arrowPosition: fixture });
      expect(toJson(renderedComponent)).toMatchSnapshot();
    });

    it('should accept "bottom-left" as value', () => {
      const fixture = 'bottom-left';
      renderedComponent.setProps({ arrowPosition: fixture });
      expect(toJson(renderedComponent)).toMatchSnapshot();
    });
  });
});
