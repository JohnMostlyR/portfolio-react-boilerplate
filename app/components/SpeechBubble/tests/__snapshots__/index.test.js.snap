// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpeechBubble /> should adopt any children 1`] = `
<Container>
  <Bubble
    arrowPosition="bottom-left"
    isGhost={false}
  >
    <span>
      A Child
    </span>
  </Bubble>
  <Arrow
    arrowPosition="bottom-left"
    isGhost={false}
  />
</Container>
`;

exports[`<SpeechBubble /> should adopt the "arrowHeight" prop 1`] = `
<Container>
  <Bubble
    arrowPosition="bottom-left"
    isGhost={false}
  />
  <Arrow
    arrowPosition="bottom-left"
    height="1vw"
    isGhost={false}
  />
</Container>
`;

exports[`<SpeechBubble /> should adopt the "backgroundColor" prop 1`] = `
<Container>
  <Bubble
    arrowPosition="bottom-left"
    backgroundColor="#000"
    isGhost={false}
  />
  <Arrow
    arrowPosition="bottom-left"
    backgroundColor="#000"
    isGhost={false}
  />
</Container>
`;

exports[`<SpeechBubble /> should adopt the "color" prop 1`] = `
<Container>
  <Bubble
    arrowPosition="bottom-left"
    color="#fff"
    isGhost={false}
  />
  <Arrow
    arrowPosition="bottom-left"
    isGhost={false}
  />
</Container>
`;

exports[`<SpeechBubble /> should adopt the "maxWidth" prop 1`] = `
<Container
  maxWidth="10em"
>
  <Bubble
    arrowPosition="bottom-left"
    isGhost={false}
  />
  <Arrow
    arrowPosition="bottom-left"
    isGhost={false}
  />
</Container>
`;

exports[`<SpeechBubble /> should adopt the "padding" prop 1`] = `
<Container>
  <Bubble
    arrowPosition="bottom-left"
    isGhost={false}
    padding="0 0px"
  />
  <Arrow
    arrowPosition="bottom-left"
    isGhost={false}
  />
</Container>
`;

exports[`<SpeechBubble /> should adopt the "showArrowBreakpoint" prop 1`] = `
<Container
  showArrowBreakpoint="100px"
>
  <Bubble
    arrowPosition="bottom-left"
    isGhost={false}
    showArrowBreakpoint="100px"
  />
  <Arrow
    arrowPosition="bottom-left"
    isGhost={false}
    showBreakpoint="100px"
  />
</Container>
`;

exports[`<SpeechBubble /> should render and match the snapshot 1`] = `
<Container>
  <Bubble
    arrowPosition="bottom-left"
    isGhost={false}
  />
  <Arrow
    arrowPosition="bottom-left"
    isGhost={false}
  />
</Container>
`;

exports[`<SpeechBubble /> the "isGhost" property should accept boolean "false" as value 1`] = `
<Container>
  <Bubble
    arrowPosition="bottom-left"
    isGhost={false}
  />
  <Arrow
    arrowPosition="bottom-left"
    isGhost={false}
  />
</Container>
`;

exports[`<SpeechBubble /> the "isGhost" property should accept boolean "true" as value 1`] = `
<Container>
  <Bubble
    arrowPosition="bottom-left"
    isGhost={true}
  />
  <Arrow
    arrowPosition="bottom-left"
    isGhost={true}
  />
</Container>
`;

exports[`<SpeechBubble /> the "makeAppear" property should accept boolean "false" as value 1`] = `
<Container
  makeAppear={false}
>
  <Bubble
    arrowPosition="bottom-left"
    isGhost={false}
  />
  <Arrow
    arrowPosition="bottom-left"
    isGhost={false}
  />
</Container>
`;

exports[`<SpeechBubble /> the "makeAppear" property should accept boolean "true" as value 1`] = `
<Container
  makeAppear={true}
>
  <Bubble
    arrowPosition="bottom-left"
    isGhost={false}
  />
  <Arrow
    arrowPosition="bottom-left"
    isGhost={false}
  />
</Container>
`;

exports[`<SpeechBubble /> the arrowPosition property should accept "bottom-left" as value 1`] = `
<Container>
  <Bubble
    arrowPosition="bottom-left"
    isGhost={false}
  />
  <Arrow
    arrowPosition="bottom-left"
    isGhost={false}
  />
</Container>
`;

exports[`<SpeechBubble /> the arrowPosition property should accept "bottom-right" as value 1`] = `
<Container>
  <Bubble
    arrowPosition="bottom-right"
    isGhost={false}
  />
  <Arrow
    arrowPosition="bottom-right"
    isGhost={false}
  />
</Container>
`;

exports[`<SpeechBubble /> the arrowPosition property should accept "top-left" as value 1`] = `
<Container>
  <Arrow
    arrowPosition="top-left"
    isGhost={false}
  />
  <Bubble
    arrowPosition="top-left"
    isGhost={false}
  />
</Container>
`;

exports[`<SpeechBubble /> the arrowPosition property should accept "top-right" as value 1`] = `
<Container>
  <Arrow
    arrowPosition="top-right"
    isGhost={false}
  />
  <Bubble
    arrowPosition="top-right"
    isGhost={false}
  />
</Container>
`;
